### ISSUE TITLE
PrimitiveType.union erases original type

### ISSUE THREAD

**Issue by asfimport** [apache/parquet-java#1456]
The following ScalaTest test case
```Java
  test("merge primitive types") {
    val expected =
      Types.buildMessage()
        .addField(
          Types
            .required(INT32)
            .as(DECIMAL)
            .precision(7)
            .scale(2)
            .named("f"))
        .named("root")

    assert(expected.union(expected) === expected)
  }
```
produces the following assertion error
```
message root {
  required int32 f;
}
 did not equal message root {
  required int32 f (DECIMAL(9,0));
}
```
This is because `PrimitiveType.union` doesn't handle original type properly. An open question is that, can two primitive types with the same primitive type name but different original types be unioned?

**Reporter**: [Cheng Lian](https://issues.apache.org/jira/secure/ViewProfile.jspa?name=lian+cheng) / @liancheng
**Assignee**: [Cheng Lian](https://issues.apache.org/jira/secure/ViewProfile.jspa?name=lian+cheng) / @liancheng
#### Related issues:
- [Release Parquet-mr 1.9.0](https://issues.apache.org/jira/browse/PARQUET-392) (blocks)
- [PrimitiveType.union accepts fixed_len_byte_array fields with different lengths when strict mode is on](https://issues.apache.org/jira/browse/PARQUET-385) (is related to)
- [GroupType.union() doesn't merge the original type](https://issues.apache.org/jira/browse/PARQUET-753) (is related to)
#### PRs and other links:
- [https://github.com/apache/parquet-mr/pull/315](https://github.com/apache/parquet-mr/pull/315)

<sub>**Note**: *This issue was originally created as [PARQUET-379](https://issues.apache.org/jira/browse/PARQUET-379). Please see the [migration documentation](https://issues.apache.org/jira/browse/PARQUET-2502) for further details.*</sub>

**Comment by asfimport** [comment]
[Cheng Lian](https://issues.apache.org/jira/browse/PARQUET-379?#comment-14933630) / @liancheng:
While trying to fix this issue, I got a problem regarding to the `strict` argument of `PrimitiveType.union`, and correspondingly, `MessageType.union`.  Seems that throughout the whole parquet-mr code base (including tests), we always call these methods with `strict` being `true`, which means schema primitive types should match.

Maybe I missed something here, but I don't see a sound use case of non-strict schema merging.  Especially, the field types of `t1.union(t2, false)` is completely determined by `t1`, rather than the "wider" ones:
```
message t1 { required int32 f; }
message t2 { required int64 f; }

t1.union(t2, false) =>
  message t3 { required int32 f; }
```
Basically we can't use such a schema to read actual Parquet files even if we add some sort of automatic "type widening" logic inside Parquet readers since the merged one above loses precision.

So my questions are:
1. Is there a practical scenario where non-strict schema merging makes sense?
1. If not, should we deprecate it? (We can't remove it since `MessageType.union(MessageType, boolean)` is part of the public API.)

**Comment by asfimport** [comment]
[Ryan Blue](https://issues.apache.org/jira/browse/PARQUET-379?#comment-14933846) / @rdblue:
I think this is part of a larger issue of handling schema evolution. The main use case I know of for union is merging file schemas into a metadata summary file. Those are no longer really needed because each schema is resolved against the requested schema individually on the reader, which eliminates the bottle-neck that the metadata file was intended to avoid. And as you note, union doesn't really create a union as one might expect: a schema that can be used to read both of the input schemas.

**Comment by asfimport** [comment]
[Cheng Lian](https://issues.apache.org/jira/browse/PARQUET-379?#comment-14934084) / @liancheng:
So deprecating non-strict schema merging seems to be reasonable? Namely, deprecate `MessageType.union(MessageType toMerge, boolean strict)`, and always set `strict` to `true` when we call this method internally.

**Comment by asfimport** [comment]
[Julien Le Dem](https://issues.apache.org/jira/browse/PARQUET-379?#comment-15062516) / @julienledem:
[~lian cheng] Did you have a patch for this.

**Comment by asfimport** [comment]
[Cheng Lian](https://issues.apache.org/jira/browse/PARQUET-379?#comment-15120370) / @liancheng:
Just opened one for this: https://github.com/apache/parquet-mr/pull/315

**Comment by asfimport** [comment]
[Julien Le Dem](https://issues.apache.org/jira/browse/PARQUET-379?#comment-15135978) / @julienledem:
 Issue resolved by pull request 315
https://github.com/apache/parquet-mr/pull/315
